Dockerfile and docker-compose

azbarcea opened the pull request:
See #1255 

It also provides a documentation how to build and use (`README.md`).

tmccormi commented:
@azbarcea you seem to have included some commits that were removed from the repo.  Please rebase and resubmit your P/R thanks!

azbarcea commented on docker/php5.6/Dockerfile:
```diff
@@ -0,0 +1,63 @@
+COPY docker-entrypoint.sh /usr/local/bin/
+
+ENTRYPOINT ["/usr/local/bin/docker-entrypoint.sh"]
+CMD ["apache2-foreground"]
```
will create 2 x `Dockerfile`s one for php5 and one for php7 ...

tmccormi replied:
we don't need php 5 at all

robbyoconnor replied:
PHP 5.x is soon to be EOL as I mentioned prior. Hence do not create one -- we don't even run it our  demo servers. As @tmccormi said, we don't need it. For informational purposes check [the following link](http://php.net/supported-versions.php). PHP 5.6 will be EOL and not supported _AT ALL_ after January 2019.

robbyoconnor replied:
I have unresolved this thread as until you provide a PHP 7.x image and nuke this one -- it's not necessarily resolved yet.

robbyoconnor replied:
Also -- another thing -- We can build from this the repo directly. We don't need to do necessarily pull from releases...we can use git tags to build releases. No need to pull from tarballs.

azbarcea replied:
will remove `php5.6` too. Would it be needed, `php7.1`, `php7.2` ? Which one would be the default?

azbarcea replied:
OK ... will use git and master branch then? The conventions is that you support and use the docker images as you use any software (releases and you can also have nightly or straight from the master).

azbarcea commented:
Default is now to `php7.0`. The reason I started with `php5.6` is that old linux distros still have `php5.6`. If not required will remove it. Here is an example were you can support multiple `php` versions. Also, we can use `apache`, `fpm`, `nginx` etc.

Later will also rebase from master.

azbarcea commented:
Testing it ... using `git` permissions are not defined for folders and files. Maybe even a `.htaccess` level for each folder (recursively) could be defined. Just a thought ...

The same thought goes into the way releases are made. During the release process, when you create the package (tarball), the same ownership policy can be defined.

robbyoconnor commented:
We're using Docker -- so we can set the distro. It doesn't matter what the host is running, the container can run whatever.

azbarcea commented:
@robbyoconnor, @tmccormi: with latest commit, I think I addressed all your concerns.

Let me know if I missed something, or you'd like something else to improve before the merge :wink:

azbarcea commented:
> @azbarcea you seem to have included some commits that were removed from the repo. Please rebase and resubmit your P/R thanks!

I have rebased and cherry-picked, forced push. It seems that 5505095e4 was replaced by b01616806bc :-)

azbarcea commented:
Anything else needed that blocks this #PR to be merged, then?

robbyoconnor commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+   librehealth:
+     depends_on:
+       - db
+     image: librehealth:${PHP_DOCKER_VERSION}-apache
```
Er see comments -- `librehealth:ehr-php7.0-apache`  -- there is little reason to support multiple PHP versions :)

robbyoconnor replied:
Also service name should be `librehealth_ehr` or `lh-ehr`, or just `ehr`.

azbarcea replied:
Will change the service name as desired. 

Which `php` versions would you prefer: `php7.0`, php7.1`, `php7.2` (`php7.3` it's on the way). I presume the current `7.0` is Ok?

robbyoconnor replied:
PHP 7.x is fine -- easy to fix.

azbarcea replied:
Fixed then.

robbyoconnor commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+
+   librehealth:
+     depends_on:
+       - db
```
Add a healthcheck. See comments on old outdated diff for resources.

azbarcea replied:
I would say that is not needed in this case. `depends_on` being sufficient.

Will add it if you prefer, though.

robbyoconnor replied:
a healthcheck is a good thing to have anyways -- lets us know when the db is not healthy

azbarcea commented:
> Is there a reason why you opted to not use a smaller image such as alpine? This is still not 100% there -- it's close though.

Yes. It helps the community identify better what it needs. With `alpine` you wont get any utility. Won't be able to install utilities. So if you want to troubleshoot or you want to use for testing or continuous integration you won't be able to use `alpine` easily. It is true that in a strong PROD environment, you'd want the image with less size and less utilities (security reasons). So, I guess in time, the community will desire multiple `php` versions, multiple webservers `nginx`/`apache`,`fpm` etc, multiple distros `debian`/`atomic`/`alpine`, multiple caching mechanisms etc. This will come with maturity.

azbarcea commented on docker/php7.0/docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+   librehealth:
+     depends_on:
+       - db
+     image: librehealth:php7.0-latest
```
Not sure what you mean. You want me to move the `docker-compose.yml` contents into the `README.md`, and later on you'll move it to a different documentation source ?

azbarcea replied:
You're not allowed to have `:` in the name of the image. After `;` is reserved for the `tag`, also used as `version` or `distro`. To address your comment, I have it as (Updated `README.md` as well):
```
REPOSITORY               TAG                 IMAGE ID            CREATED             SIZE
librehealth_ehr          php7.1-apache       9b51d050d73c        8 minutes ago       742MB
```

azbarcea commented on docker/php7.0/docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+       MYSQL_USER: libreehr
+       MYSQL_PASSWORD: s3cret
+
+   librehealth:
```
Will do.

azbarcea commented on docker/php7.0/docker-entrypoint.sh:
```diff
@@ -0,0 +1,95 @@
+			--directory /usr/src/librehealth \
+			--owner "$user" --group "$group" \
+			. | tar --extract --file -
+		echo >&2 "Complete! LibreHealth has been successfully copied to $PWD"
```
Will do.

azbarcea replied:
Fixed with 4c2498e

robbyoconnor commented on docker/php7.0/docker-entrypoint.sh:
```diff
@@ -0,0 +1,95 @@
+	if [ ! -e index.php ] && [ ! -e version.php ]; then
+		echo >&2 "LibreHealth not found in $PWD - copying now..."
+		if [ "$(ls -A)" ]; then
+			echo >&2 "WARNING: $PWD is not empty - press Ctrl+C now if this is an error!"
```
This won't be run in an interactive setting...

azbarcea replied:
removed ...

robbyoconnor commented on docker/php7.0/docker-entrypoint.sh:
```diff
@@ -0,0 +1,95 @@
+			--directory /usr/src/librehealth \
+			--owner "$user" --group "$group" \
+			. | tar --extract --file -
+		echo >&2 "Complete! LibreHealth has been successfully copied to $PWD"
```
```shell
	echo >&2 "Complete! LibreHealth EHR has been successfully copied to $PWD"
```

azbarcea replied:
Fixed. Removed `/usr/src` per previous request

robbyoconnor commented on docker/php7.0/docker-entrypoint.sh:
```diff
@@ -0,0 +1,95 @@
+			echo >&2 "WARNING: $PWD is not empty - press Ctrl+C now if this is an error!"
+			( set -x; ls -A; sleep 10 )
+		fi
+		tar --create \
```
Why?

azbarcea replied:
removed.

tmccormi commented:
@robbyoconnor  please review (again) and test this if you can.

tmccormi commented:
There was a request for some data for alex to use for testing, etc.   I have pushed the nhanes test database to the repo:  

- commit a7fde7b349af10fc2e34607a6e1cb5f69611587
- sql/libreehr_nhanes.sql.gz

robbyoconnor commented:
@tmccormi -- none of my review comments were addressed.

azbarcea commented on docker/php7.0/Dockerfile:
```diff
@@ -0,0 +1,63 @@
+	rm librehealth.tar.gz; \
+	mv /usr/src/lh-ehr-${LIBREHEALTH_VERSION}* /usr/src/librehealth; \
+	chown -R www-data:www-data /usr/src/librehealth
+
```
fixed with 4c2498e

azbarcea commented on docker/php7.0/docker-entrypoint.sh:
```diff
@@ -0,0 +1,95 @@
+	fi
+
+	if [ ! -e index.php ] && [ ! -e version.php ]; then
+		echo >&2 "LibreHealth not found in $PWD - copying now..."
```
Fixed with 35476a445

azbarcea commented:
Hi @robbyoconnor, @tmccormi - all your review comments were addressed. Tested! Please verify.

azbarcea commented on docker/php7.0/docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+       MYSQL_DATABASE: libreehr
+       MYSQL_USER: libreehr
+       MYSQL_PASSWORD: s3cret
+
```
Fixed. Snippet from `docker logs`:
```
MYSQL host db is unavailable - sleeping ...
MYSQL host db is unavailable - sleeping ...
MYSQL host db is unavailable - sleeping ...
MYSQL host db is unavailable - sleeping ...
MYSQL host db is unavailable - sleeping ...
MYSQL host db is unavailable - sleeping ...
MYSQL host db is unavailable - sleeping ...
MYSQL host db is unavailable - sleeping ...
MYSQL host db is unavailable - sleeping ...
MYSQL host db is unavailable - sleeping ...
mysqld is alive
DB is up - continue ...
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 192.168.32.3. Set the 'ServerName' directive globally to suppress this message
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 192.168.32.3. Set the 'ServerName' directive globally to suppress this message
[Thu Oct 25 00:56:15.814218 2018] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.25 (Debian) PHP/7.0.32 configured -- resuming normal operations
[Thu Oct 25 00:56:15.814260 2018] [core:notice] [pid 1] AH00094: Command line: 'apache2 -D FOREGROUND'
```

robbyoconnor commented:
@azbarcea -- I just wanted to let ya know I didn't forget -- I will review this tomorrow (Monday).

hzbarcea commented:
Are all the comments addressed? Can this be merged?

robbyoconnor commented:
> Are all the comments addressed? Can this be merged?

Not yet. I still have to test it. The comments thus far have been things that I saw that were immediate deal-breakers for me.

robbyoconnor commented:
Also not a deal breaker but we almost universally use nginx on everything _EXCEPT_ for EHR...how willing would you be to make an image for that @azbarcea --- for that I'd be willing put in some work myself if you're not willing :)

tmccormi commented:
> Also not a deal breaker but we almost universally use nginx on everything _EXCEPT_ for EHR...how willing would you be to make an image for that @azbarcea --- for that I'd be willing put in some work myself if you're not willing :)

Later .... not this time says tony.  We've never tested the EHR in nGinX

robbyoconnor commented:
> > Also not a deal breaker but we almost universally use nginx on everything _EXCEPT_ for EHR...how willing would you be to make an image for that @azbarcea --- for that I'd be willing put in some work myself if you're not willing :)
> 
> Later .... not this time says tony. We've never tested the EHR in nGinX

Was just a suggestion :) Going to test this in a bit.

robbyoconnor commented:
> I use it for other things. Just needs testing and some updates to the docs to be a viable option I'm sure.

Willing to lend a hand if you want.

robbyoconnor commented on docker/README.md:
```diff
@@ -0,0 +1,98 @@
+
+    $ export PHP_DOCKER_VERSION=php7.1
+    $ cd lh-ehr/docker
+    $ docker build --no-cache -t librehealth_ehr:${PHP_DOCKER_VERSION}-apache .
```
No need for this.

robbyoconnor commented on docker/README.md:
```diff
@@ -0,0 +1,98 @@
+    $ export PHP_DOCKER_VERSION=php7.1
+    $ cd lh-ehr/docker
+    $ docker build --no-cache -t librehealth_ehr:${PHP_DOCKER_VERSION}-apache .
+    $ docker-compose up -p librehealth -d
```
This will build it if it doesn't exist.

robbyoconnor commented on docker/README.md:
```diff
@@ -0,0 +1,98 @@
+    $ docker build --no-cache -t librehealth_ehr:${PHP_DOCKER_VERSION}-apache .
+    $ docker-compose up -p librehealth -d
+
+### 3. Start LibreHealth composition
```
```markdown 
### 3. Start LibreHealth EHR container
```

azbarcea replied:
Fixed in 7f335bab6

robbyoconnor commented on docker/README.md:
```diff
@@ -0,0 +1,98 @@
+### 3. Start LibreHealth composition
+
+    $ docker-compose up -p librehealth -d
+
```
Actually this whole thing can go.

azbarcea replied:
Fixed in 7f335bab6

robbyoconnor commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+       LIBREHEALTH_DB_USER: libreehr
+       LIBREHEALTH_DB_PASSWORD: s3cret
+volumes:
+    db_data:
```
```yaml
version: '3.3'

services:
   db:
     image: mysql:5.7
     volumes:
       - db_data:/var/lib/mysql
     restart: always
     environment:
       MYSQL_ROOT_PASSWORD: m4ster_s3cret
       MYSQL_DATABASE: libreehr
       MYSQL_USER: libreehr
       MYSQL_PASSWORD: s3cret
   librehealth_ehr:
     depends_on:
       - db
     # image: librehealth_ehr:${PHP_DOCKER_VERSION-php7.1}-apache
     content: ..
     ports:
       - "8000:80"
     restart: always
     environment:
       LIBREHEALTH_DB_HOST: db
       LIBREHEALTH_DB_PORT: 3306
       LIBREHEALTH_DB_NAME: libreehr
       LIBREHEALTH_DB_USER: libreehr
       LIBREHEALTH_DB_PASSWORD: s3cret
volumes:
    db_data:
```

Let's use this for now -- until we push the image. Let the image be built from the Dockerfile.

robbyoconnor commented:
@azbarcea ignore my comments and just merge this PR and it should be good to go.

robbyoconnor commented:
@azbarcea ignore my comments and just merge this PR and it should be good to go: https://github.com/azbarcea/lh-ehr/pull/1

azbarcea commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,28 @@
+       LIBREHEALTH_DB_USER: libreehr
+       LIBREHEALTH_DB_PASSWORD: s3cret
+volumes:
+    db_data:
```
Not sure what you need here ...

azbarcea commented:
@azbarcea, I don't think I have the appropriate entitlements to merge myself.

> Another thing move env variables to a .env file -- this provides isolation and all secrets go in one place.

It would require 2 `.env` files as you can see the `ENV` is not quite the same. Do I understand you'd prefer `db.env` and `librehealth_ehr.env`. Considering the size of the `docker-compose.yml` (28 lines) I consider adding another 2 files would make the whole composition more complex. Higher barrier for the community developers to use/develop the `docker-compose.yml`, unless you add the `*.env` files to `.gitignore`.

robbyoconnor commented on docker/README.md:
```diff
@@ -17,7 +17,7 @@ Build using your desired `php` version, by picking from the following: `php7.0-a
     $ docker-compose up -p librehealth -d
 
-### 3. Start LibreHealth composition
+### 3. Start LibreHealth EHR composition
```
Still wrong.

robbyoconnor commented:
@azbarcea the pull request is against your repository. Once you make the changes you'll see I did exactly what you said. So just merge it and your changes will appear here: https://github.com/azbarcea/lh-ehr/pull/1/files

Notice how it's in your repo against your PR branch

robbyoconnor commented:
Here's a diff you can apply: https://patch-diff.githubusercontent.com/raw/azbarcea/lh-ehr/pull/1.patch

robbyoconnor commented:
@azbarcea you're good to go -- I made minor tweaks and squashed your commits -- I left the changes I made as a separate commit.

robbyoconnor commented:
@azbarcea @tmccormi this is good to go. I added initial NHANES sample data -- which makes initial container creation for mysql slow but not much we can do there. This is good in my eyes.
